design/reboot-interface.md

```diff
@@ -0,0 +1,155 @@
+time may still fall before some portion of the future-dated requests. Such
+requests will be treated as if they occurred at the present time rather than at
+the future time given in the timestamp.
+
+A new ``pendingRebootSince`` field in the status will contain the timestamp of
+the earliest reboot request following the last booted time, if there has been
+any and the machine is provisioned.
+
+Client code may be assured that the node has been fenced when any of the
+following are true:
+
+* The Host has been deprovisioned (in this case the ``lastBooted`` field will
+  be empty)
+* The ``lastBooted`` time is after the time of the reboot request
+* The ``poweredOn`` status is ``false``
+* The ``pendingRebootSince`` time is empty (checking this is necessary only if
+  fairly severe clock skew is possible)
+
+## Drawbacks
+
+The need for the consumer to supply a recent timestamp means that the design
+requires some co-operation from the client that is not naturally enforced by
+the API. Without good documentation this could lead to confusion and incorrect
+usage. It is also vulnerable to clock skew in severe cases.
+
+## Alternatives
+
+The fencing code could perform the reboot in an imperative, rather than
+declarative, manner by manipulating the ``Online`` spec of the Host in sequence
+to ``false`` and then back to ``true`` again once the ``poweredOn`` status was
+seen to be ``false``. However, in the presence of multiple actors this approach
+is prone to race conditions or controller fights.
+
+A separate ``HostRebootRequest`` CRD would allow maintaining a record of past
+reboot requests at the Host level in addition to those created at the Machine
+level by the fencing code. The question of the interface between this new CRD
+and the BareMetalHost would be identical to that between the Machine-level CRD
+and the BareMetalHost in the design above.
+
+A separate HostRebootRequest CRD could remain robust to multiple actors in the
+absence of a reboot request interface in the BareMetalHost itself if control
+were inverted. That is to say, rather than the HostRebootRequest acting on the
+BareMetalHost by imperatively manipulating the ``Online`` spec, the
+BareMetalHost would search for pending reboot requests naming it as the target
+and initiate a reboot if any are found. This would mean that multiple requests
+would be consolidated into a single reboot at the point of action rather than
+at the point of receiving each request. This would maintain a Host-level record
+of reboot requests. It would also allow for future-dated requests if we still
+allow the user to provide a timestamp, or could prevent the ambiguity by using
+the creation timestamp of the resource (which is not under the control of the
+user) instead. However, this would be considerably more involved to implement,
+and may become inefficient (requiring large amounts of data to be transferred
+from etcd) if stale requests are allowed to accumulate.
```

Comment:
@mhrivnak WDYT about this last paragraph as a compromise? It's more work to write/maintain, but it should maintain correctness without sacrificing the UX.

Reply:
Presumably all matching reboot request CRs would be updated with status, so that the clients that created them could watch them?

Reply:
Yes. I'm assuming that the controller for the RebootRequest would watch the BareMetalHost and update its status once it saw the Host reboot.

Reply:
If the baremetal operator is supposed to respond to the separate reboot request, I would expect it to be the one updating the CR with status.

Reply:
Is anything allowed to update the status (as opposed to the spec) of another resource? I'm not clear on what the rules are there.

Reply:
Good question. Technically it's possible. It does seem to go against the usual pattern of having the "owning" controller responsible for the status fields, though. It also seems like mixing management of the request CR and host CR in one controller is going to complicate that controller.

What exactly are the conditions under which we have a race when there are multiple requests to reboot the same host?

If 2 reboot CRs are created at the same time, their controller will have a race to turn the host off. But when the second CR is processed I would expect the controller to see that the host is already in a state where the BMO is trying to turn it off, and immediately update its status to reflect that the reboot is in progress. Then when the host state continues to change, we may also have a race to turn it back on with the same sort of results. So, that's not a problem.

If one reboot CR is already being processed with the host in an off state and being turned back on when a second reboot CR is created, I would expect the controller to recognize that, interpret the state as not needing a reboot (because the host was powered off to begin with), and update its status to indicate that the host is coming back up. So that's also not a problem.

If one reboot CR is being processed with the host in an on state after a reboot but the reboot CR does not yet reflect that when a second CR is created, I would expect a second reboot to be triggered by the new CR. The first CR may, or may not, be set to indicate that the host was powered back on before the second reboot, but I'm not sure it matters very much. The controller should update it as the second CR is processed. At worst it's slightly inaccurate information about when the host was powered on and if the reboots come back-to-back like that it's not _that_ inaccurate.

Am I missing a case?

Reply:
Seems reasonable to me. Usually for any resource, there should be one "owning" controller that makes changes to the world based on the spec and writes to the status. I don't see a problem in this case with having one controller owning both resources, since the HRR is basically a child of the BMH.